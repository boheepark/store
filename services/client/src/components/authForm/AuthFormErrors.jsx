import React from 'react';

import './AuthFormErrors.css';


const AuthFormErrors = (props) => {
  return (
    <div>
      <ul className="validation-list">
        {
          props.rules.map((rule) => {
            return (
              <li
                className={rule.valid ? "success" : "error"}
                key={rule.id}
              >{rule.message}
              </li>
            );
          })
        }
      </ul>
    </div>
  );
};

export default AuthFormErrors;
