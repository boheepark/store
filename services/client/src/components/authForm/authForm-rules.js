export const signupRules = [
  {
    id: 1,
    field: 'username',
    message: 'Username must be greater than 5 characters.',
    valid: false
  },
  {
    id: 2,
    field: 'email',
    message: 'Email must be greater than 5 characters.',
    valid: false
  },
  {
    id: 3,
    field: 'email',
    message: 'Email must be a valid email address.',
    valid: false
  },
  {
    id: 4,
    field: 'password',
    message: 'Password must be greater than 10 characters.',
    valid: false
  }
];

export const signinRules = [
  {
    id: 1,
    field: 'email',
    message: 'Email is required.',
    valid: false
  },
  {
    id: 2,
    field: 'password',
    message: 'Password is required.',
    valid: false
  }
];
