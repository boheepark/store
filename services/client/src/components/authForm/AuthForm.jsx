import React, {Component} from 'react';
import axios from 'axios';
import {Redirect} from 'react-router-dom';
import {signupRules, signinRules} from './authForm-rules';
import AuthFormErrors from "./AuthFormErrors";


class AuthForm extends Component {

  constructor(props) {
    super(props);
    this.state = {
      data: {
        username: '',
        email: '',
        password: ''
      },
      signupRules: signupRules,
      signinRules: signinRules,
      valid: false
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  };

  componentDidMount() {
    this.clearForm();
  };

  componentWillReceiveProps(nextProps) {
    if (this.props.form !== nextProps.form) {
      this.clearForm();
      this.resetRules();
      // this.initRules();
    }
  };

  handleChange(event) {
    const data = this.state.data;
    data[event.target.name] = event.target.value;
    this.setState(data);
    this.validateForm();
  };

  handleSubmit(event) {
    event.preventDefault();
    const form = this.props.form;
    let data;
    if (form === 'signin') {
      data = {
        email: this.state.data.email,
        password: this.state.data.password
      };
    } else if (form === 'signup') {
      data = {
        username: this.state.data.username,
        email: this.state.data.email,
        password: this.state.data.password
      };
    }
    const url = `${process.env.REACT_APP_USERS_SERVICE_URL}/auth/${form}`;
    axios.post(url, data)
    .then((res) => {
      this.clearForm();
      this.props.signin(res.data.data.token);
    })
    .catch((err) => {
      console.log(err);
      this.props.createMessage(`${form} failed.`, 'danger');
      // if(form === 'signin') {
      //   this.props.createMessage('Signin failed.', 'danger');
      // }
      // if(form === 'signup') {
      //   this.props.createMessage('Signup failed.', 'danger');
      // }
    });
  };

  clearForm() {
    this.setState({
      data: {
        username: '',
        email: '',
        password: ''
      }
    });
  };


  allTrue() {
    let rules;
    if(this.props.form === 'signup') {
      rules = signupRules;
    } else if(this.props.form === 'signin') {
      rules = signinRules;
    }
    for (const rule of rules) {
      if (!rule.valid) {
        return false;
      }
    }
    return true;
  };

  // initRules() {
  //   const rules = this.state.rules;
  //   for (const rule of rules) {
  //     rule.valid = false;
  //   }
  //   this.setState({
  //     rules: rules
  //   });
  // };

  resetRules() {
    if(this.props.form === 'signin') {
      const rules = this.state.signinRules;
      for(const rule of rules) {
        rule.valid = false;
      }
      this.setState({
        signinRules: rules
      });
    } else if(this.props.form === 'signup') {
      const rules = this.state.signupRules;
      for(const rule of rules) {
        rule.valid = false;
      }
      this.setState({
        signupRules: rules
      });
    }
    this.setState({
      valid: false
    });
  };

  validateEmail(email) {
    // eslint-disable-next-line
    let re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return re.test(email);
  };

  validateForm() {
    const data = this.state.data;

    this.resetRules();

    if(this.props.form === 'signup') {
      const rules = this.state.signupRules;
      if(data.username.length > 5) {
        rules[0].valid = true;
      }
      if(data.email.length > 5) {
        rules[1].valid = true;
      }
      if(this.validateEmail(data.email)) {
        rules[2].valid = true;
      }
      if(data.password.length > 10) {
        rules[3].valid = true;
      }
      this.setState({
        signupRules: rules
      });
      if(this.allTrue()) {
        this.setState({
          valid: true
        });
      }
    } else if(this.props.form === 'signin') {
      const rules = this.state.signinRules;
      if(data.email.length > 0) {
        rules[0].valid = true;
      }
      if(data.password.length > 0) {
        rules[1].valid = true;
      }
      this.setState({
        signinRules: rules
      });
      if(this.allTrue()) {
        this.setState({
          valid: true
        });
      }
    }
  };

  render() {
    if (this.props.isAuthenticated) {
      return <Redirect to='/'/>;
    }
    let rules;
    if(this.props.form === 'signup') {
      rules = this.state.signupRules;
    } else if (this.props.form === 'signin') {
      rules = this.state.signinRules;
    }
    return (
      <div>
        <h1 style={{'textTransform': 'capitalize'}}>{this.props.form}</h1>
        <hr/>
        <br/>
        <AuthFormErrors
          form={this.props.form}
          rules={rules}
        />
        <form onSubmit={(event) => this.handleSubmit(event)}>
          {
            this.props.form === 'signup' &&
            <div className="form-group">
              <input
                name="username"
                className="form-control input-lg"
                type="text"
                placeholder="Enter a username"
                required
                value={this.state.data.username}
                onChange={this.handleChange}
              />
            </div>
          }
          <div className="form-group">
            <input
              name="email"
              className="form-control input-lg"
              type="email"
              placeholder="Enter an email address"
              required
              value={this.state.data.email}
              onChange={this.handleChange}
            />
          </div>
          <div className="form-group">
            <input
              name="password"
              className="form-control input-lg"
              type="password"
              placeholder="Enter a password"
              required
              value={this.state.data.password}
              onChange={this.handleChange}
            />
          </div>
          <input
            type="submit"
            className="btn btn-secondary btn-lg btn-block"
            value="Submit"
            disabled={!this.state.valid}
          />
        </form>
      </div>
    );
  };
};

export default AuthForm;
