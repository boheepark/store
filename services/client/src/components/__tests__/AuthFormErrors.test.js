import React from 'react';
import {shallow, mount} from 'enzyme';
import renderer from 'react-test-renderer';
import {MemoryRouter as Router} from 'react-router-dom';

import AuthFormErrors from '../authForm/AuthFormErrors';
import {signupRules, signinRules} from '../authForm/authForm-rules';


const testData = [
  {
    form: 'signup',
    rules: signupRules
  },
  {
    form: 'signin',
    rules: signinRules
  }
];


describe('AuthFormErrors tests', () => {
  testData.forEach((el) => {
    const component = <AuthFormErrors {...el} />;

    it(`${el.form} form errors render properly`, () => {
      const wrapper = shallow(component);
      const ul = wrapper.find('ul');
      expect(ul.length).toBe(1);
      const li = wrapper.find('li');
      expect(li.length).toBe(el.rules.length);
      el.rules.forEach((rule, i) => {
        expect(li.get(i).props.children).toContain(
          rule.message
        );
      });
    });

    it(`${el.form} form errors render a snapshot properly`, () => {
      const tree = renderer.create(
        <Router>
          <AuthFormErrors {...el} />
        </Router>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
