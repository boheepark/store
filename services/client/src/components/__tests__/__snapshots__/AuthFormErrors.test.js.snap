// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthFormErrors tests signin form errors render a snapshot properly 1`] = `
<div>
  <ul
    className="validation-list"
  >
    <li
      className="error"
    >
      Email is required.
    </li>
    <li
      className="error"
    >
      Password is required.
    </li>
  </ul>
</div>
`;

exports[`AuthFormErrors tests signup form errors render a snapshot properly 1`] = `
<div>
  <ul
    className="validation-list"
  >
    <li
      className="error"
    >
      Username must be greater than 5 characters.
    </li>
    <li
      className="error"
    >
      Email must be greater than 5 characters.
    </li>
    <li
      className="error"
    >
      Email must be a valid email address.
    </li>
    <li
      className="error"
    >
      Password must be greater than 10 characters.
    </li>
  </ul>
</div>
`;
